taichi/backends/struct_opengl.cpp

```diff
@@ -0,0 +1,91 @@
+#include "struct_opengl.h"
+
+TLANG_NAMESPACE_BEGIN
+namespace opengl {
+
+OpenglStructCompiler::CompiledResult OpenglStructCompiler::run(SNode &node)
+{
+  TI_ASSERT(node.type == SNodeType::root);
+  collect_snodes(node);
+  // The host side has run this!
+  // infer_snode_properties(node);
+
+  auto snodes_rev = snodes_;
+  std::reverse(snodes_rev.begin(), snodes_rev.end());
+
+  for (auto &n : snodes_rev) {
+    generate_types(*n);
+  }
+  CompiledResult result;
+  result.source_code = std::move(src_code_);
+  result.root_size = compute_snode_size(node);
+  return result;
+}
+
+void OpenglStructCompiler::collect_snodes(SNode &snode) {
+  snodes_.push_back(&snode);
+  for (int ch_id = 0; ch_id < (int)snode.ch.size(); ch_id++) {
+    auto &ch = snode.ch[ch_id];
+    collect_snodes(*ch);
+  }
+}
+// TODO(archibate): really need fit struct_metal so much?
```

Comment:
Feel free to use what you think is the most convenient layout ;)

FYI, Metal's design follows the legacy memory layout in https://github.com/taichi-dev/taichi/blob/master/misc/obsolete_code/struct.h#L366. This layout also matches how LLVM generates those structs, so we can leverage unified memory.

Reply:
Agree, GLSL don't have pointers, we could use just `mem[tmp233]` without `S1_get0(tmp233)` making output so complicated. And `struct_opengl.cpp` is no longer needed since all can be done by array indexing, will work on this in later stages to prevent merge conflicts.